File: QuantLib-SWIG/Ruby/test/capfloor.rb
=begin
 Copyright (C) 2000, 2001, 2002, 2003 RiskMap srl

 This file is part of QuantLib, a free-software/open-source library
 for financial quantitative analysts and developers - http://quantlib.org/

 QuantLib is free software: you can redistribute it and/or modify it under the
 terms of the QuantLib license.  You should have received a copy of the
 license along with this program; if not, please email ferdinando@ametrano.net
 The license is also available online at http://quantlib.org/html/license.html

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the license for more details.
=end

# $Id$

require 'QuantLib'
require 'test/unit'
require 'test/unit/ui/console/testrunner'

class CapFloorTest < Test::Unit::TestCase
  include QuantLib
  def name
    case @method_name
      when 'testStrikeDependency'
        "Testing cap/floor dependency on strike"
      when 'testConsistency'
        "Testing consistency between cap, floor and collar"
      when 'testParity'
        "Testing put/call parity for cap and floor"
      when 'testCachedValue'
        "Testing cap/floor value against cached values"
    end
  end
  def setup
    @termStructure = TermStructureHandle.new
    @nominals = [100.0]
    @rollingConvention = 'modifiedFollowing'
    @frequency = 2
    @index = Xibor.new('Euribor',12/@frequency,'Months',
                       @termStructure)
    @calendar = @index.calendar
    @settlementDays = 2
    @today = @calendar.roll(Date.todaysDate)
    @settlement = @calendar.advance(@today,
                                    @settlementDays,"days",
                                    'following')
    @fixingDays = 2
    @termStructure.linkTo!(FlatForward.new(@today,@settlement,
                                           0.05,DayCounter.new('Act/360')))
  end
  def makeLeg(startDate,length)
    endDate = @calendar.advance(startDate,length,"years",
                                @rollingConvention)
    FloatingRateCouponVector(@nominals,startDate,endDate,@frequency,@calendar,
                             @rollingConvention,@index,@fixingDays)
  end
  def makeCapFloor(kind,leg,strike,volatility)
    kind.new(leg,[strike],@termStructure,makeEngine(volatility))
  end
  def makeEngine(volatility)
    BlackCapFloorEngine.new(
      BlackModel.new(
        MarketElementHandle.new(SimpleMarketElement.new(volatility)),
        @termStructure))
  end
  def testStrikeDependency
    startDate = @termStructure.referenceDate
    [1, 2, 3, 5, 7, 10, 15, 20].each do |length|
      [0.01, 0.05, 0.10, 0.15, 0.20].each do |vol|
        [Cap,Floor].each do |kind|

          strikes = [0.03, 0.04, 0.05, 0.06, 0.07]
          values = strikes.map { |s|
            leg = makeLeg(startDate,length)
            instrument = makeCapFloor(kind,leg,s,vol)
            instrument.NPV
          }

          if kind == Cap
            # NPV must decrease with strike
            (1...values.length).each do |i|
              unless values[i] <= values[i-1]
                flunk(<<-MESSAGE

    NPV is increasing with the strike in a cap:
        length: #{length} years
        volatility: #{vol}
        value: #{values[i-1]} at strike: #{strikes[i-1]*100}%
        value: #{values[i]  } at strike: #{strikes[i]*100}%

                      MESSAGE
                      )
              end
            end
          else
            # NPV must increase with strike
            (1...values.length).each do |i|
              unless values[i] >= values[i-1]
                flunk(<<-MESSAGE

    NPV is decreasing with the strike in a cap:
        length: #{length} years
        volatility: #{vol}
        value: #{values[i-1]} at strike: #{strikes[i-1]*100}%
        value: #{values[i]  } at strike: #{strikes[i]*100}%

                      MESSAGE
                      )
              end
            end
          end
        end
      end
    end
  end
  def testConsistency
    startDate = @termStructure.referenceDate
    [1, 2, 3, 5, 7, 10, 15, 20].each do |length|
      [0.03, 0.04, 0.05, 0.06, 0.07].each do |capRate|
        [0.03, 0.04, 0.05, 0.06, 0.07].each do |floorRate|
          [0.01, 0.05, 0.10, 0.15, 0.20].each do |vol|

            leg = makeLeg(startDate,length)
            cap = makeCapFloor(Cap,leg,capRate,vol)
            floor = makeCapFloor(Floor,leg,floorRate,vol)
            collar = Collar.new(leg,[capRate],[floorRate],
                                @termStructure,makeEngine(vol))

            unless ((cap.NPV-floor.NPV) - collar.NPV).abs <= 1.0e-10
              flunk(<<-MESSAGE

    inconsistency between cap, floor and collar:
        length      : #{length} years
        volatility  : #{vol}
        cap value   : #{cap.NPV  } at strike: #{capRate*100}%
        floor value : #{floor.NPV} at strike: #{floorRate*100}%
        collar value: #{collar.NPV}

                    MESSAGE
                    )
            end
          end
        end
      end
    end
  end
  def testParity
    startDate = @termStructure.referenceDate
    [1, 2, 3, 5, 7, 10, 15, 20].each do |length|
      [0.03, 0.04, 0.05, 0.06, 0.07].each do |strike|
        [0.01, 0.05, 0.10, 0.15, 0.20].each do |vol|
          
          leg = makeLeg(startDate,length)
          cap = makeCapFloor(Cap,leg,strike,vol)
          floor = makeCapFloor(Floor,leg,strike,vol)
          swap = SimpleSwap.new(true,startDate,length,"years",@calendar,
                                @rollingConvention,@nominals[0],
                                @frequency,strike,@index.isAdjusted?,
                                @index.dayCounter,@frequency,
                                @index,@fixingDays,0.0,@termStructure)
          unless ((cap.NPV-floor.NPV) - swap.NPV).abs <= 1.0e-10
            flunk(<<-MESSAGE

    put/call parity violated:
        length     : #{length} years
        volatility : #{vol}
        strike     : #{strike*100}%
        cap value  : #{cap.NPV}
        floor value: #{floor.NPV}
        swap value : #{swap.NPV}

                  MESSAGE
                  )
          end
        end
      end
    end
  end
  def testCachedValue
    cachedToday = Date.new(14,3,2002)
    cachedSettlement = Date.new(18,3,2002)
    @termStructure.linkTo!(FlatForward.new(cachedToday,cachedSettlement,
                                           0.05, DayCounter.new('Act/360')))
    startDate = @termStructure.referenceDate
    leg = makeLeg(startDate,20)
    cap = makeCapFloor(Cap,leg,0.07,0.20)
    floor = makeCapFloor(Floor,leg,0.03,0.20)
    cachedCapNPV = 6.958278080775
    cachedFloorNPV = 2.700476857631

    unless (cap.NPV-cachedCapNPV).abs <= 1.0e-11
      flunk(<<-MESSAGE

    failed to reproduce cached cap value:
        calculated: #{cap.NPV}
        expected:   #{cachedCapNPV}

            MESSAGE
            )
    end
    unless (floor.NPV-cachedFloorNPV).abs <= 1.0e-11
      flunk(<<-MESSAGE

    failed to reproduce cached floor value:
        calculated: #{floor.NPV}
        expected:   #{cachedFloorNPV}

            MESSAGE
            )
    end
  end
end

if $0 == __FILE__
  Test::Unit::UI::Console::TestRunner.run(CapFloorTest)
end

File: QuantLib-SWIG/Python/QuantLib/test/capfloor.py
"""
 Copyright (C) 2000, 2001, 2002, 2003 RiskMap srl

 This file is part of QuantLib, a free-software/open-source library
 for financial quantitative analysts and developers - http://quantlib.org/

 QuantLib is free software: you can redistribute it and/or modify it under the
 terms of the QuantLib license.  You should have received a copy of the
 license along with this program; if not, please email ferdinando@ametrano.net
 The license is also available online at http://quantlib.org/html/license.html

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the license for more details.
"""

from QuantLib import *
import unittest

class CapFloorTest(unittest.TestCase):
    def setUp(self):
        self.termStructure = TermStructureHandle()
        self.nominals = [100.0]
        self.rollingConvention = 'modifiedFollowing'
        self.frequency = 2
        self.index = Xibor('Euribor',12/self.frequency,'Months',
                           self.termStructure)
        self.calendar = self.index.calendar()
        self.settlementDays = 2
        self.today = self.calendar.roll(Date_todaysDate())
        self.settlement = self.calendar.advance(self.today,
                                                self.settlementDays,"days",
                                               'following')
        self.fixingDays = 2
        self.termStructure.linkTo(FlatForward(self.today, self.settlement,
                                              0.05, DayCounter('Act/360')))
    def makeLeg(self,startDate,length):
        endDate = self.calendar.advance(startDate,length,"years",
                                        self.rollingConvention)
        return FloatingRateCouponVector(self.nominals,startDate,endDate,
                                        self.frequency,self.calendar,
                                        self.rollingConvention,
                                        self.index, self.fixingDays)
    def makeCapFloor(self,kind,leg,strike,volatility):
        return kind(leg,[strike],
                    self.termStructure,
                    self.makeEngine(volatility))
    def makeEngine(self,volatility):
        return BlackCapFloorEngine(
                   BlackModel(
                       MarketElementHandle(
                           SimpleMarketElement(volatility)),
                       self.termStructure))
    # check 1
    def testStrikeDependency(self):
        "Testing cap/floor dependency on strike"
        cases = [ (length,vol,kind)
                   for length in [1, 2, 3, 5, 7, 10, 15, 20]
                   for vol in [0.01, 0.05, 0.10, 0.15, 0.20]
                   for kind in [Cap,Floor] ]
        startDate = self.termStructure.referenceDate()
        for (length,vol,kind) in cases:
            strikes = [0.03, 0.04, 0.05, 0.06, 0.07]
            values = []
            for s in strikes:
                leg = self.makeLeg(startDate,length)
                instrument = self.makeCapFloor(kind,leg,s,vol)
                values.append(instrument.NPV())

            if kind == Cap:
                # NPV must decrease with strike
                for i in range(1,len(values)):
                    if values[i] > values[i-1]:
                        self.fail("""
NPV is increasing with the strike in a cap:
    length: %d years
    volatility: %8.4f
    value: %8.4f at strike: %8.4f
    value: %8.4f at strike: %8.4f
                                  """ % (length,vol,
                                         values[i-1],strikes[i-1]*100,
                                         values[i],  strikes[i]*100))
            else:
                # NPV must increase with strike
                for i in range(1,len(values)):
                    if values[i] < values[i-1]:
                        self.fail("""
NPV is decreasing with the strike in a floor:
    length: %d years
    volatility: %8.4f
    value: %8.4f at strike: %8.4f
    value: %8.4f at strike: %8.4f
                                  """ % (length,vol,
                                         values[i-1],strikes[i-1]*100,
                                         values[i],  strikes[i]*100))
    # check 2
    def testConsistency(self):
        "Testing consistency between cap, floor and collar"
        cases = [ (length,capRate,floorRate,vol)
                   for length in [1, 2, 3, 5, 7, 10, 15, 20]
                   for capRate in [0.03, 0.04, 0.05, 0.06, 0.07]
                   for floorRate in [0.03, 0.04, 0.05, 0.06, 0.07]
                   for vol in [0.01, 0.05, 0.10, 0.15, 0.20] ]
        startDate = self.termStructure.referenceDate()
        for (length,capRate,floorRate,vol) in cases:
            leg = self.makeLeg(startDate,length)
            cap = self.makeCapFloor(Cap,leg,capRate,vol)
            floor = self.makeCapFloor(Floor,leg,floorRate,vol)
            collar = Collar(leg,[capRate],[floorRate],
                            self.termStructure,
                            self.makeEngine(vol))
            if abs((cap.NPV()-floor.NPV()) - collar.NPV()) > 1.0e-10:
                self.fail("""
inconsistency between cap, floor and collar:
    length      : %d years
    volatility  : %8.4f
    cap value   : %8.4f at strike: %8.4f
    floor value : %8.4f at strike: %8.4f
    collar value: %8.4f
                          """ % (length,vol,
                                 cap.NPV(),capRate*100,
                                 floor.NPV(),floorRate*100),
                                 collar.NPV())
    # check 3
    def testParity(self):
        "Testing put/call parity for cap and floor"
        cases = [ (length,strike,vol)
                   for length in [1, 2, 3, 5, 7, 10, 15, 20]
                   for strike in [0.03, 0.04, 0.05, 0.06, 0.07]
                   for vol in [0.01, 0.05, 0.10, 0.15, 0.20] ]
        startDate = self.termStructure.referenceDate()
        for (length,strike,vol) in cases:
            leg = self.makeLeg(startDate,length)
            cap = self.makeCapFloor(Cap,leg,strike,vol)
            floor = self.makeCapFloor(Floor,leg,strike,vol)
            swap = SimpleSwap(1,startDate,length,"years",self.calendar,
                              self.rollingConvention,self.nominals[0],
                              self.frequency,strike,self.index.isAdjusted(),
                              self.index.dayCounter(),self.frequency,
                              self.index,self.fixingDays,0.0,
                              self.termStructure)
            if abs((cap.NPV()-floor.NPV()) - swap.NPV()) > 1.0e-10:
                self.fail("""
put/call parity violated:
    length     : %d years
    volatility : %8.4f
    strike     : %8.4f
    cap value  : %8.4f
    floor value: %8.4f
    swap value : %8.4f
                          """ % (length,vol,
                                 strike*100,cap.NPV(),
                                 floor.NPV(),swap.NPV()))
    # check 4
    def testCachedValue(self):
        "Testing cap/floor value against cached values"
        cachedToday = Date(14,3,2002)
        cachedSettlement = Date(18,3,2002)
        self.termStructure.linkTo(FlatForward(cachedToday, cachedSettlement,
                                              0.05, DayCounter('Act/360')))
        startDate = self.termStructure.referenceDate()
        leg = self.makeLeg(startDate,20)
        cap = self.makeCapFloor(Cap,leg,0.07,0.20)
        floor = self.makeCapFloor(Floor,leg,0.03,0.20)
        cachedCapNPV = 6.960233718984
        cachedFloorNPV = 2.701296290808
        if abs(cap.NPV()-cachedCapNPV) > 1.0e-11:
            self.fail("""
failed to reproduce cached cap value:
    calculated: %18.12f
    expected:   %18.12f
                      """ % (cap.NPV(),cachedCapNPV))
        if abs(floor.NPV()-cachedFloorNPV) > 1.0e-11:
            self.fail("""
failed to reproduce cached floor value:
    calculated: %18.12f
    expected:   %18.12f
                      """ % (floor.NPV(),cachedFloorNPV))


if __name__ == '__main__':
    import QuantLib
    print 'testing QuantLib', QuantLib.__version__
    suite = unittest.TestSuite()
    suite.addTest(unittest.makeSuite(CapFloorTest,'test'))
    unittest.TextTestRunner(verbosity=2).run(suite)
    raw_input('press return to continue')
